Logger

ASalvail opened the pull request.

ASalvail committed:
Introduced a logger interface with a default in-memory implementation.

ASalvail committed:
Refactored MonitorVariable and removed obsolete tasks.

ASalvail committed:
Added task/views to interact with loggers.

ASalvail committed:
Changed default values of frequency.

ASalvail committed:
Added tests

ASalvail committed:
Fixed tests for them to pass. Will open an issue to find a more permanent solution.

ASalvail committed:
deleted old logger.py module

MarcCote commented on tests/optimizers/tests_optimizers.py:
```diff
@@ -101,10 +101,10 @@ def test_adagrad():
-        gparam = tasks.MonitorVariable(loss.gradients[param])
-        trainer.append_task(gparam)
+        gparam = views.MonitorVariable(loss.gradients[param])
+        trainer._graph_updates.update(gparam.updates)
```
We could do this instead

``` python
logger = tasks.Logger(gparam)
trainer.append_task(logger)
```

ASalvail replied:
We'd need a logger that doesn't keep an history.

MarcCote replied:
I like the `Tracker` idea except it is somehow redundant having to create a `MonitorVariable` first and then pass it to a `Tracker`. But I guess there is no magic, we have to it to the trainer so it can be updated.

ASalvail replied:
Another thing we could do is pass a variable directly, then the `Tracker` would create the view itself.

``` python
def __init__(self, *vars, **freqs):
  super().__init__(*[views.MonitorVariable(v) for v in vars], **freqs)
```

MarcCote replied:
But you'd probably want to access that monitored variable somehow, like you do below: `gparam.value`. If the tracker creates it, we would need to add a `value` property, like the old TrackVariable task had, so we can do `my_tracker.value`. Still we would need two lines, one to create the Tracker object and another to add it to the trainer.

ASalvail replied:
No, we'd use `tracker[0]`.

MarcCote replied:
ok

MarcCote commented on tests/optimizers/tests_optimizers.py:
```diff
@@ -101,10 +101,10 @@ def test_adagrad():
         # After 30 epochs, param should be around the center and gradients near 0.
         assert_array_almost_equal(param.get_value(), center)
-        assert_array_almost_equal(gparam.value, 0.)
+        assert_array_almost_equal(gparam.var.get_value(), 0.)
```
Then use this:
`assert_array_almost_equal(logger[-1][0], 0.)`

I think the logger usage could be improve, it doesn't seem convivial at all.

ASalvail replied:
Which part? And what would you suggest?

MarcCote commented on tests/optimizers/tests_optimizers.py:
```diff
@@ -70,14 +70,14 @@ def test_sgd():
     # Since the problem is well-conditionned and we use an optimal gradient step 1/L,
     # two epochs should be enough for `param` to be around `center` and the gradients near 0.
     assert_array_almost_equal(param.get_value(), center, decimal=6)
-    assert_array_almost_equal(gparam.value, 0.)
```
Same as below

MarcCote commented on tests/optimizers/tests_optimizers.py:
```diff
@@ -70,14 +70,14 @@ def test_sgd():
     # Monitor the gradient of `loss` w.r.t. to `param`.
-    gparam = tasks.MonitorVariable(loss.gradients[param])
-    trainer.append_task(gparam)
+    gparam = views.MonitorVariable(loss.gradients[param])
```
Same as below (in the `test_adagrad` function)

ASalvail committed:
Added a tracking task which takes the role of MonitorVariable.

ASalvail commented:
The new task fixes #50 .
However, in the test, we still talk with the view instead of the task. What would you suggest to make the logging interface better? I agress that `logger[0][0]` is a bit klunky...

ASalvail committed:
Updated how an item is accessed with a single list element in the `_history`.

ASalvail commented:
Klunky calls are fixed (partially) in 4fa113159a16474f204f9880664ff1afeb71cdef.

MarcCote commented:
@ASalvail what do you mean by partially? You mean the logger is still klunky? We could add a test in `Logger` and if there is only one view, do not return tuple but return a single value instead.

ASalvail commented:
The problem with this is that now the interface would be different depending on the call.
Can we use a starred tuple as a return argument? ie `return *self._history`

MarcCote commented:
Ah yes, I hadn't thought of that.

MarcCote commented:
I don't have any more comments, is this ready to merge?

ASalvail committed:
Changed expected arguments of `Tracker` for variables instead of views.

Had to put AverageTrainingError in tasks to prevent import cycle.
Had to change order of tests to make sure the tracker is executed before the stopping criteria fires.

ASalvail commented:
Now it is.

MarcCote commented:
Oups, should I'd merged this PR before #51 ? Can you rebase master and I'll merge it.

ASalvail committed:
Fixed test after merging.

ASalvail commented:
Both were based on master. The rebase and some corrections are done.
